### PR TITLE
[aptos-vm] Add a feature flag to disallow user published natives

### PR DESCRIPTION
**Pull request by runtian-zhou**
## Description

Disallowing users to publish modules with natives.

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [x] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?

Added test cases for both positive and negative cases.

## Key Areas to Review
N/A

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: third_party/move/move-core/types/src/vm_status.rs [hunk 1]
```diff
@@ -721,8 +721,8 @@ pub enum StatusCode {
     DEPENDENCY_LIMIT_REACHED = 1124,
     // This error indicates that unstable bytecode generated by the compiler cannot be published to mainnet
     UNSTABLE_BYTECODE_REJECTED = 1125,
+    USER_DEFINED_NATIVE = 1126,
```
Comment (vgao1996) [inline third_party/move/move-core/types/src/vm_status.rs:724]: Hmm... you might need to rebase your PR. 1126 is already taken by a new status code PROGRAM_TOO_COMPLEX

File: aptos-move/aptos-vm/src/aptos_vm.rs [hunk 2]
```diff
@@ -1509,6 +1523,12 @@ impl AptosVM {
                 .is_enabled(FeatureFlag::SAFER_RESOURCE_GROUPS),
         )?;
         verifier::event_validation::validate_module_events(session, modules)?;
+        if self
+            .features()
+            .is_enabled(FeatureFlag::DISALLOW_USER_NATIVES)
+        {
+            verifier::native_validation::validate_module_natives(modules)?;
+        }
```
Comment (vgao1996) [inline aptos-move/aptos-vm/src/aptos_vm.rs:1531]: Wonder if we should move this check to a slightly earlier point, considering how trivial it is. No string opinions though

File: third_party/move/move-core/types/src/vm_status.rs (hunk 1 above)
Comment (vgao1996) [reply third_party/move/move-core/types/src/vm_status.rs:724]: Also wondering if we should change the native to something more descriptive, like, USER_NATIVE_NOT_ALLOWED?

File: aptos-move/aptos-vm/src/aptos_vm.rs (hunk 2 above)
Comment (runtian-zhou) [reply aptos-move/aptos-vm/src/aptos_vm.rs]: Done!

File: third_party/move/move-core/types/src/vm_status.rs (hunk 1 above)
Comment (runtian-zhou) [reply third_party/move/move-core/types/src/vm_status.rs]: Done!

**Review by vgao1996 — APPROVED** [review]

File: aptos-move/aptos-vm/src/verifier/native_validation.rs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+/// Validate that only system address can publish new natives.
+pub(crate) fn validate_module_natives(modules: &[CompiledModule]) -> VMResult<()> {
+    for module in modules {
+        let module_address = module.self_addr();
+        for def in module.function_defs() {
+            if def.is_native() {
+                if !module_address.is_special() {
+                    return Err(
+                        PartialVMError::new(StatusCode::USER_DEFINED_NATIVE_NOT_ALLOWED)
+                            .with_message("Cannot publish native function".to_string())
```
Comment (georgemitenkov) [inline aptos-move/aptos-vm/src/verifier/native_validation.rs]: nit: "Cannot publish native function to non-special address"?

File: aptos-move/aptos-vm/src/verifier/native_validation.rs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    CompiledModule,
+};
+use move_core_types::vm_status::StatusCode;
+
+/// Validate that only system address can publish new natives.
+pub(crate) fn validate_module_natives(modules: &[CompiledModule]) -> VMResult<()> {
+    for module in modules {
+        let module_address = module.self_addr();
+        for def in module.function_defs() {
+            if def.is_native() {
```
Comment (georgemitenkov) [inline aptos-move/aptos-vm/src/verifier/native_validation.rs]: nit: combine two ifs into one to have less indent?

File: aptos-move/aptos-vm/src/aptos_vm.rs [hunk 5]
```diff
@@ -759,6 +759,22 @@ impl AptosVM {
...
+        // Native entry function is forbidden.
+        if self
+            .features()
+            .is_enabled(FeatureFlag::DISALLOW_USER_NATIVES)
+            && function.is_native()
+        {
+            return Err(
+                PartialVMError::new(StatusCode::USER_DEFINED_NATIVE_NOT_ALLOWED)
+                    .with_message(
+                        "Execute user defined native entry function is not allowed".to_string(),
```
Comment (georgemitenkov) [inline aptos-move/aptos-vm/src/aptos_vm.rs:771]: nit: Execute  --> Executing

**Review by georgemitenkov — APPROVED** [review]
What about

**Pull request merged**